openforcefield/tests/test_molecule.py

```diff
@@ -529,6 +529,109 @@ def test_name(self):
         molecule.name = name
         assert molecule.name == name
 
+    def test_hill_formula(self):
+        """Test that making the hill formula is consistent between input methods and ordering"""
+        # make sure smiles match reference
+        molecule_smiles = Molecule.from_smiles('CCO')
+        assert molecule_smiles.hill_formula == 'C2H6O1'
+        # make sure is not order dependent
+        molecule_smiles_reverse = Molecule.from_smiles('OCC')
+        assert molecule_smiles.hill_formula == molecule_smiles_reverse.hill_formula
+        # make sure files and smiles match
+        molecule_file = Molecule.from_file(get_data_file_path('molecules/ethanol.mol2'))
+        assert molecule_smiles.hill_formula == molecule_file.hill_formula
+        # make sure the topology molecule gives the same formula
+        from openforcefield.topology.topology import TopologyMolecule, Topology
+        topology = Topology.from_molecules(molecule_smiles)
+        topmol = TopologyMolecule(molecule_smiles, topology)
+        assert molecule_smiles.hill_formula == Molecule.to_hill_formula(topmol)
+        # make sure the networkx matches
+        assert molecule_smiles.hill_formula == Molecule.to_hill_formula(molecule_smiles.to_networkx())
+
+    def test_isomorphic(self):
+        """Test the different levels of isomorphic matching"""
+        # check that hill formula fails are caught
+        ethanol = Molecule.from_smiles('CCO')
+        acetaldehyde = Molecule.from_smiles('CC=O')
```

Comment:
The behavior of `from_smiles` is completely delegated to the cheminformatics toolkits. It just so happens that the current implementation of their `from_smiles` functions preserves the order of the heavy atoms from the SMILES in this case, but that may not always be true. If the behavior changes in the future (or if another toolkit is added that doesn't behave this way), these tests will fail, and it'll take us a while to figure out why.

In these sorts of cases, I recommend making a test function to create a molecule using OFFTK's API. This may be paranoid, but it could save us a big headache in the future. Here's an example: https://github.com/openforcefield/openforcefield/blob/b8b7b314602a2685d5fe9441d01bb0a7f7c8badc/openforcefield/tests/test_forcefield.py#L259-L315

For now, I'd put the `create_acetaldehyde` function in test_forcefields, and just import it. It's sloppy, but that will keep them all in one place for when we consolidate them later.

Reply:
Thanks for pointing this out I have added in the functions to the test_ff file just need to remember to remove `create_reversed_ethanol()` function.